org.osgi.test.common/src/main/java/org/osgi/test/common/context/CloseableBundleContext.java

```diff
@@ -132,14 +134,31 @@ public void close() {
 		bundles.stream()
 			.filter(installed)
 			.forEach(uninstallBundle);
-		services.forEach(this::ungetService);
+		bundles.clear();
```

Comment:
> What is the reasoning behind calling `clear()` on all the state structures?

Because is it necessary for correct operation. :-) The closeable resource for services is closed _after_ the closeable resource for bundle context. So when the service trackers called BundleContext.ungetService, the services map was still fully populated with services and use counts and this then affected the parent closeable bundle context. Clearing is the state thing would be in if the test case properly cleans up after itself, so we must do the same to allow the closeable resources to be closed in any order.

Its is not a gc issue, so setting to null is only likely to lead to NPEs.

Reply:
> > What is the reasoning behind calling `clear()` on all the state structures?
> 
> The closeable resource for services is closed _after_ the closeable resource for bundle context. So when the service trackers called BundleContext.ungetService, the services map was still fully populated with services and use counts and this then affected the parent closeable bundle context. Clearing is the state things would be in if the test case properly cleans up after itself, so we must do the same to allow the closeable resources to be closed in any order.
> 

To my mind, it seems wrong that we allow methods on `CloseableBundleContext` to be called after it has been `close()`d. You can't write to a file after it has been closed, and you can't use a service after it has been "ungot". I think we should even go as far as to say that we should throw `IllegalStateException` in such cases to explicitly notify the developer of their error.

I think this is the reason why I didn't rely on the `CloseableResource` auto-closing feature in `BundleContext`, as in doing so you relinquish control over the shutdown order. As I understand it, Jupiter guarantees that the extensions' "afterEach()" methods are called in reverse order of the "beforeEach()" methods (which is the order that the extensions are registered/declared). If `ServiceExtension` explicitly cleaned up after itself rather than relying on `CloseableResource` functionality to do it, then the ordering problem goes away (you could still allow `BundleContextExtension` to fall back to `CloseableResource` auto-cleanup in this case, as that will come last).

> Its is not a gc issue, so setting to null is only likely to lead to NPEs.

It will only cause NPEs if you attempt to call any of the methods after `close()` has been called, which I feel should be illegal anyway.

But if we persist down this path, then setting the state collections to `Collections.emptyMap()` may be more efficient than calling `clear()` on each one.

Reply:
> As I understand it, Jupiter guarantees that the extensions' "afterEach()" methods are called in reverse order of the "beforeEach()" methods (which is the order that the extensions are registered/declared).

Yes, _IF_ the extension is applied with `@ExtendWith`. But I think we want to be able to use `BundleContextExtension.getBundleContext` from ServiceExtension and ConfigurationExtension (and any additional extensions we author in the future) and we want to be able to use `ServiceExtension.getServiceConfiguration` from ConfigurationExtension (and any additional extensions we author in the future) without requiring all test writers to say `@ExtendWith({BundleContextExtension.class, ServiceExtension.class})` to use ServiceExtension or to say `@ExtendWith({BundleContextExtension.class, ServiceExtension.class, ConfigurationExtension.class})` to use ConfigurationExtension. I think that is a real usability issue. So CloseableResources are nice here. If there is some way to control the ordering of closing CloseableResources, that would be great. But otherwise, we need to be able to deal with then closing in any order.

> then setting the state collections to `Collections.emptyMap()` may be more efficient than calling `clear()` on each one.

I am not sure why you are so concerned about performance here. I would expect that a test case will almost never use more than a few services. So the set/map to be cleared is quite small. I mean we are not talking thousands of entries in the set/map.

> It will only cause NPEs if you attempt to call any of the methods after `close()` has been called, which I feel should be illegal anyway.

These are closeable to service our implementation needs of cleaning up resources used by the execution context. The closeable nature is not visible to any one else but our implementation. Users cannot close them. So we don't need to enforce any don't-use-after-close semantics.

Reply:
> > As I understand it, Jupiter guarantees that the extensions' "afterEach()" methods are called in reverse order of the "beforeEach()" methods (which is the order that the extensions are registered/declared).
> 
> Yes, _IF_ the extension is applied with `@ExtendWith`. But I think we want to be able to use `BundleContextExtension.getBundleContext` from ServiceExtension and ConfigurationExtension (and any additional extensions we author in the future) and we want to be able to use `ServiceExtension.getServiceConfiguration` from ConfigurationExtension (and any additional extensions we author in the future) without requiring all test writers to say `@ExtendWith({BundleContextExtension.class, ServiceExtension.class})` to use ServiceExtension or to say `@ExtendWith({BundleContextExtension.class, ServiceExtension.class, ConfigurationExtension.class})` to use ConfigurationExtension. I think that is a real usability issue.

I fully agree with this. Any solution should not require you to explicitly declare the extension's dependent extensions and should seamlessly interoperate with user code that directly uses the extension.

> So CloseableResources are nice here. If there is some way to control the ordering of closing CloseableResources, that would be great. But otherwise, we need to be able to deal with them closing in any order.
>

Much as this feels a bit wrong, I can't think of a better way at the moment.
 
> > then setting the state collections to `Collections.emptyMap()` may be more efficient than calling `clear()` on each one.
> 
> I am not sure why you are so concerned about performance here. 
> 

Performance was only a secondary concern. My primary concern was the correctness of allowing methods to be called on a closed object.

Reply:
So I think we are ok with the current state of things? I do appreciate the robust conversation here!

Reply:
Thanks for indulging me! Aside from the minor performance movement of using emptyMap() (non critical), if you should choose, I'm happy with the way it is.

Reply:
> If there is some way to control the ordering of closing CloseableResources, that would be great.

To close the loop on this, I found and comment on https://github.com/junit-team/junit5/issues/2211